src/pages/domains/[domain]/index.tsx

```diff
@@ -1,26 +1,148 @@
 
-export default function CreateNimiPage() {
-  const router = useRouter();
-  const { domain } = router.query;
+interface NimiSnapshot {
+  publisher: string;
+  cid: string | null;
+  nimi: Nimi;
+  createdAt: string;
+  updatedAt: string;
+  id: string;
+}
+
+export async function fetchNimiIPNS(ensName: string) {
+  const url = `http://127.0.0.1:4002/ens/has-nimi-ipns?domain=${ensName}`;
+
+  const rawResponse = await fetch(url, {
+    method: 'GET',
+  }).then();
 
-  // DOn't render anything on server side
-  if (typeof window === 'undefined') {
-    console.log('Server side rendering');
-    return null;
+  if (!rawResponse.ok) {
+    throw new Error('Network response was not ok');
   }
 
+  const data = (await rawResponse.json()) as {
+    data: {
+      ipns?: string;
+      nimi?: NimiSnapshot;
+    };
+  };
+
+  return data.data;
+}
+
+export async function generateNimiFromENS(ensName: string) {
+  const { data } = await nimiClient.get<{
+    data: {
+      nimi: Nimi;
+    };
+  }>(`/nimi/generate`, {
+    params: {
+      domain: ensName,
+    },
+  });
+
+  return data.data;
+}
+
+interface CreateNimiPageProps {
+  availableThemes: ReturnType<typeof getAvailableThemesByPOAPs>;
+  domain: string;
+  ipnsKey?: string;
+  nimiSnapshot?: NimiSnapshot;
+  initialNimi: Nimi;
+  poapList: POAPToken[];
+}
+
+// This is the page that will be server side rendered
+export async function getServerSideProps(
+  context: GetServerSidePropsContext<{
+    domain: string;
+  }>
+): Promise<{
+  props: CreateNimiPageProps;
+}> {
+  let props = {
+    availableThemes: [],
+    domain: context?.params?.domain,
+    initialNimi: {} as Nimi,
+    poapList: [],
+  } as CreateNimiPageProps;
+
+  try {
+    if (!context?.params?.domain || context?.params?.domain === '[object Object]') {
+      throw new Error('No domain provided');
+    }
+
+    const domain = context.params.domain;
+    // Nimi Snapshot from the API
+    const nimiIPNSSnapshot = await fetchNimiIPNS(domain);
+    // If the user has a Nimi IPNS, redirect to the Nimi page
+    if (nimiIPNSSnapshot.ipns && nimiIPNSSnapshot.nimi) {
+      props = {
+        ...props,
+        ipnsKey: nimiIPNSSnapshot.ipns,
+        nimiSnapshot: nimiIPNSSnapshot.nimi,
+        initialNimi: nimiIPNSSnapshot.nimi.nimi,
+      };
+    } else {
+      const { nimi } = await generateNimiFromENS(domain);
+      props['initialNimi'] = nimi;
+    }
+
+    if (!props.initialNimi) {
+      throw new Error('No Nimi found');
+    }
+
+    // Attempt to load
+    const poapList = await fetchUserPOAPs(props.initialNimi.ensAddress as string);
+    const availableThemes = getAvailableThemesByPOAPs(poapList);
+
+    if (poapList) {
+      props = {
+        ...props,
+        poapList,
+        availableThemes,
+      };
+    }
+  } catch (e) {
+    console.log(e);
+  }
+  // If the user doesn't have a Nimi IPNS, generate a Nimi from the ENS name
+  return { props };
+}
+
```

Comment:
there is still a way to use reactQuery with ssr here is article  https://tanstack.com/query/latest/docs/react/guides/ssr?from=reactQueryV3&original=https%3A%2F%2Freact-query-v3.tanstack.com%2Fguides%2Fssr#using-initialdata